Kernel/Storage/SD/Registers.h

```diff
@@ -0,0 +1,200 @@
+ */
+
+#pragma once
+
+#include <AK/Endian.h>
+#include <AK/Types.h>
+
+namespace Kernel::SD {
+
+enum class HostVersion : u8 {
+    Version1,
+    Version2,
+    Version3,
+    Unknown
+};
+
+// SD Host Controller Simplified Specification Version 3.00
+// NOTE: The registers must be 32 bits, because of a quirk in the RPI.
+struct SDRegisters {
+    u32 argument_2;
+    u32 block_size_and_block_count;
+    u32 argument_1;
+    u32 transfer_mode_and_command;
+    u32 response_0;
+    u32 response_1;
+    u32 response_2;
+    u32 response_3;
+    u32 buffer_data_port;
+    u32 present_state;
+    u32 host_configuration_0;
+    u32 host_configuration_1;
+    u32 interrupt_status;
+    u32 interrupt_status_enable;
+    u32 interrupt_signal_enable;
+    u32 host_configuration_2;
+    u32 capabilities_0;
+    u32 capabilities_1;
+    u32 maximum_current_capabilities;
+    u32 maximum_current_capabilities_reserved;
+    u32 force_event_for_auto_cmd_error_status;
+    u32 adma_error_status;
+    u32 adma_system_address[2];
+    u32 preset_value[4];
+    u32 reserved_0[28];
+    u32 shared_bus_control;
+    u32 reserved_1[6];
+    struct [[gnu::packed]] {
+        u8 interrupt_signal_for_each_slot;
+        u8 : 8;
+        HostVersion specification_version_number;
+        u8 vendor_version_number;
+    } slot_interrupt_status_and_version;
+};
+static_assert(AssertSize<SDRegisters, 256>());
+
+union OperatingConditionRegister {
+    u32 raw;
+    struct {
+        u32 : 15;
+        u32 vdd_voltage_window_27_28 : 1;
+        u32 vdd_voltage_window_28_29 : 1;
+        u32 vdd_voltage_window_29_30 : 1;
+        u32 vdd_voltage_window_30_31 : 1;
+        u32 vdd_voltage_window_31_32 : 1;
+        u32 vdd_voltage_window_32_33 : 1;
+        u32 vdd_voltage_window_33_34 : 1;
+        u32 vdd_voltage_window_34_35 : 1;
+        u32 vdd_voltage_window_35_36 : 1;
+        u32 switching_to_18v_accepted : 1;
+        u32 : 2;
+        u32 over_2tb_support_status : 1;
+        u32 : 1;
+        u32 uhs2_card_status : 1;
+        u32 card_capacity_status : 1;
+        u32 card_power_up_status : 1;
+    };
+};
+static_assert(AssertSize<OperatingConditionRegister, 4>());
+
+// CID response: "20 bits of the response data (R[127:8]) stored in the Response register at REP[119:0]"
+union CardIdentificationRegister {
+    u32 raw[4];
+
+    struct [[gnu::packed]] {
+        u32 manufacturing_date : 12;
+        u32 : 4;
+        u32 product_serial_number : 32;
+        u32 product_revision : 8;
+        u64 product_name : 40;
+        u32 oem_id : 16;
+        u32 manufacturer_id : 8;
+    };
+};
+static_assert(AssertSize<CardIdentificationRegister, 16>());
+
+union SDConfigurationRegister {
+    u8 raw[8];
+
+    struct {
+        u32 scr_structure : 4;
+        u32 sd_specification : 4;
+        u32 data_status_after_erase : 1;
+        u32 sd_security : 3;
+        u32 sd_bus_widths : 4;
+        u32 sd_specification3 : 1;
+        u32 extended_security : 4;
```

Comment:
I really don't like this. Bit-fields are fine for data you want to pack but for MMIO access this is certainly not acceptable, especially when we do want to be compatible with big endian CPUs (although we support non currently, it's a good practice).

Reply:
Please use bit-wise operations as we use those all over the kernel codebase instead.

Reply:
As far as I can tell this seems to be a struct locate in MMIO, but copied out of and then accessed

Reply:
If the enidanness becomes a problem,
GCC seems to have `__attribute__ ((scalar_storage_order("little-endian")))`, not sure about clang